common/parallel/TreeCrawler.go

```diff
@@ -156,11 +163,26 @@ func (c *crawler) processOneDirectory(ctx context.Context) (bool, error) {
 	// finally, update shared state (inside the lock)
 	c.cond.L.Lock()
 	defer c.cond.L.Unlock()
-	for _, d := range foundDirectories {
-		c.unstartedDirs <- d
+
+	c.unstartedDirs = append(c.unstartedDirs, foundDirectories...) // do NOT try to wait here if unstartedDirs is getting big. May cause deadlocks, due to all workers waiting and none processing the queue
+	c.dirInProgressCount--                                         // we were doing something, and now we have finished it
+	c.cond.Broadcast()                                             // let other workers know that the state has changed
+
+	// If our queue of unstarted stuff is getting really huge,
+	// reduce our parallelism in the hope of preventing further excessive RAM growth.
+	// (It's impossible to know exactly what to do here, because we don't know whether more workers would _clear_
+	// the queue more quickly; or _add to_ the queue more quickly.  It depends on whether the directories we process
+	// next contain mostly child directories or if they are "leaf" directories containing mostly just files.  But,
+	// if we slowly reduce parallelism the end state is roughly equivalent to a single-threaded depth-first traversal, which
+	// is generally fine in terms of memory usage on most folder structures)
+	const maxQueueDirectories = 1000 * 1000
+	shouldShutSelfDown := len(c.unstartedDirs) > maxQueueDirectories && // we are getting way too much stuff queued up
+		workerIndex > (c.parallelism/4) && // never shut down the last ones, since we need something left to clear the queue
+		time.Since(c.lastAutoShutdown) > time.Second // adjust slightly gradually
+	if shouldShutSelfDown {
```

Comment:
Just a thought, but would only working on the newest items help as well?

I mean, the thinking is that we are more likely to clear a single branch (and probably remove a bunch of items from the pool) if we prefer only newer items, rather than just older ones, since we're perpetually opening a whole new, likely larger can of worms with every lower-level branch we open.

Reply:
This is only really a thought since we now have the grace of being able to implement this as a double ended queue